File: vss-web/app/src/simulatorSession.ts
import type { Catalog, Demonstration, Preset, Value } from './types'

export type EyeMode = 'left' | 'both' | 'right'
type Eye = Exclude<EyeMode, 'both'>

export interface SessionLayer {
  selectedDemonstrations: Record<string, string>
  manual: Record<string, Value>
  maskedFallback: Record<string, Value>
  maskedByBoth: Record<string, true>
}

export interface SimulatorSession {
  eyeMode: EyeMode
  layers: Record<EyeMode, SessionLayer>
}

const emptyLayer = (): SessionLayer => ({
  selectedDemonstrations: {},
  manual: {},
  maskedFallback: {},
  maskedByBoth: {},
})

export function createSession(hasEyeSpecificContent = false): SimulatorSession {
  return {
    eyeMode: hasEyeSpecificContent ? 'both' : 'left',
    layers: { left: emptyLayer(), both: emptyLayer(), right: emptyLayer() },
  }
}

export function setEyeMode(session: SimulatorSession, eyeMode: EyeMode): SimulatorSession {
  return { ...session, eyeMode }
}

export function currentLayer(session: SimulatorSession): SessionLayer {
  return session.layers[session.eyeMode]
}

const preset = (catalog: Catalog, id: string) => catalog.presets.find(item => item.id === id)
const intrinsic = (value: Preset) => Object.keys(value.left).length > 0 || Object.keys(value.right).length > 0

function presetValues(value: Preset, target: EyeMode): Record<string, Value> {
  if (Object.keys(value.both).length > 0) return value.both
  return target === 'both' ? {} : value[target]
}

function layerPresets(layer: SessionLayer, catalog: Catalog): string[] {
  return [...new Set(catalog.articles.flatMap(article =>
    article.demonstrations.find(item => item.id === layer.selectedDemonstrations[article.id])?.presets ?? [],
  ))]
}

export function activePresets(session: SimulatorSession, catalog: Catalog): string[] {
  if (session.eyeMode !== 'both') return layerPresets(session.layers[session.eyeMode], catalog)
  const shared = layerPresets(session.layers.both, catalog)
  const authored = [...layerPresets(session.layers.left, catalog), ...layerPresets(session.layers.right, catalog)]
    .filter(id => preset(catalog, id) && intrinsic(preset(catalog, id)!))
  return [...new Set([...shared, ...authored])]
}

export function selectedDemonstration(session: SimulatorSession, articleId: string): string | undefined {
  if (session.eyeMode !== 'both') return session.layers[session.eyeMode].selectedDemonstrations[articleId]
  return session.layers.both.selectedDemonstrations[articleId]
    ?? session.layers.left.selectedDemonstrations[articleId]
    ?? session.layers.right.selectedDemonstrations[articleId]
}

function demonstrationValues(catalog: Catalog, demonstration: Demonstration, target: EyeMode) {
  return Object.fromEntries(demonstration.presets.flatMap(id => Object.entries(
    preset(catalog, id) ? presetValues(preset(catalog, id)!, target) : {},
  )))
}

function demonstrationsConflict(catalog: Catalog, selected: Demonstration, other: Demonstration, target: EyeMode): boolean {
  const values = demonstrationValues(catalog, selected, target)
  return Object.entries(demonstrationValues(catalog, other, target))
    .some(([setting, value]) => setting in values && !valuesEqual(values[setting], value))
}

function valuesEqual(left: Value, right: Value): boolean {
  return Array.isArray(left) && Array.isArray(right)
    ? left[0] === right[0] && left[1] === right[1]
    : left === right
}

function transitionLayer(
  layer: SessionLayer,
  catalog: Catalog,
  target: EyeMode,
  articleId: string,
  selected: Demonstration,
  enabled: boolean,
): SessionLayer {
  const before = layerPresets(layer, catalog)
  const selections = { ...layer.selectedDemonstrations }
  if (!enabled) {
    delete selections[articleId]
  } else {
    for (const article of catalog.articles) {
      const other = article.demonstrations.find(item => item.id === selections[article.id])
      if (article.id !== articleId && other && demonstrationsConflict(catalog, selected, other, target)) {
        delete selections[article.id]
      }
    }
    selections[articleId] = selected.id
  }
  const next = { ...layer, selectedDemonstrations: selections }
  return cascadePresetTransition(catalog, target, before, layerPresets(next, catalog), next)
}

function settingsForPresets(catalog: Catalog, target: EyeMode, presets: string[]) {
  return new Set(presets.flatMap(id => {
    const value = preset(catalog, id)
    return value ? Object.keys(presetValues(value, target)) : []
  }))
}

function cascadePresetTransition(
  catalog: Catalog,
  target: EyeMode,
  before: string[],
  after: string[],
  layer: SessionLayer,
): SessionLayer {
  const beforeSettings = settingsForPresets(catalog, target, before)
  const afterSettings = settingsForPresets(catalog, target, after)
  const manual = { ...layer.manual }
  const maskedFallback = { ...layer.maskedFallback }
  for (const setting of new Set([...beforeSettings, ...afterSettings])) {
    if (afterSettings.has(setting)) {
      if (setting in manual) {
        maskedFallback[setting] = manual[setting]
        delete manual[setting]
      }
    } else if (beforeSettings.has(setting)) {
      if (setting in manual) delete maskedFallback[setting]
      else if (setting in maskedFallback) {
        manual[setting] = maskedFallback[setting]
        delete maskedFallback[setting]
      }
    }
  }
  return { ...layer, manual, maskedFallback }
}

function updateLayer(session: SimulatorSession, target: EyeMode, layer: SessionLayer): SimulatorSession {
  return { ...session, layers: { ...session.layers, [target]: layer } }
}

function maskSettings(session: SimulatorSession, targets: Eye[], settings: Iterable<string>, masked: boolean) {
  let next = session
  for (const target of targets) {
    const layer = next.layers[target]
    const maskedByBoth = { ...layer.maskedByBoth }
    for (const setting of settings) {
      if (masked) maskedByBoth[setting] = true
      else delete maskedByBoth[setting]
    }
    next = updateLayer(next, target, { ...layer, maskedByBoth })
  }
  return next
}

function sharedSettings(session: SimulatorSession, catalog: Catalog) {
  return new Set([
    ...Object.keys(session.layers.both.manual),
    ...settingsForPresets(catalog, 'both', layerPresets(session.layers.both, catalog)),
  ])
}

export function selectDemonstration(
  session: SimulatorSession,
  catalog: Catalog,
  articleId: string,
  demonstrationId: string,
): SimulatorSession {
  const article = catalog.articles.find(item => item.id === articleId)
  const selected = article?.demonstrations.find(item => item.id === demonstrationId)
  if (!article || !selected) return session
  const presets = selected.presets.map(id => preset(catalog, id)).filter((item): item is Preset => Boolean(item))
  const isIntrinsic = presets.some(intrinsic)
  const targets: EyeMode[] = isIntrinsic
    ? (['left', 'right'] as Eye[]).filter(target => presets.some(item => Object.keys(item[target]).length > 0))
    : [session.eyeMode]
  const enabled = !targets.every(target => session.layers[target].selectedDemonstrations[articleId] === demonstrationId)
  let next = session
  for (const target of targets) {
    next = updateLayer(next, target, transitionLayer(next.layers[target], catalog, target, articleId, selected, enabled))
    if (target !== 'both' && enabled) {
      next = maskSettings(next, [target], Object.keys(demonstrationValues(catalog, selected, target)), false)
    }
  }
  if (targets.includes('both')) {
    const changed = Object.keys(demonstrationValues(catalog, selected, 'both'))
    if (enabled) next = maskSettings(next, ['left', 'right'], changed, true)
    else {
      const owned = sharedSettings(next, catalog)
      next = maskSettings(next, ['left', 'right'], changed.filter(setting => !owned.has(setting)), false)
    }
  }
  return isIntrinsic && enabled ? { ...next, eyeMode: 'both' } : next
}

export function editSetting(session: SimulatorSession, settingId: string, value: Value): SimulatorSession {
  const target = session.eyeMode
  const layer = session.layers[target]
  let next = updateLayer(session, target, { ...layer, manual: { ...layer.manual, [settingId]: value } })
  return target === 'both'
    ? maskSettings(next, ['left', 'right'], [settingId], true)
    : maskSettings(next, [target], [settingId], false)
}

export function resetSetting(session: SimulatorSession, settingId: string, catalog?: Catalog): SimulatorSession {
  const target = session.eyeMode
  const layer = session.layers[target]
  const manual = { ...layer.manual }
  delete manual[settingId]
  let next = updateLayer(session, target, { ...layer, manual })
  if (target === 'both' && (!catalog || !sharedSettings(next, catalog).has(settingId))) {
    next = maskSettings(next, ['left', 'right'], [settingId], false)
  }
  return next
}

function applyLayer(result: Record<string, Value>, layer: SessionLayer, catalog: Catalog, target: EyeMode) {
  for (const item of catalog.presets) {
    if (!layerPresets(layer, catalog).includes(item.id)) continue
    for (const [setting, value] of Object.entries(presetValues(item, target))) {
      if (!(setting in layer.maskedByBoth)) result[setting] = value
    }
  }
  for (const [setting, value] of Object.entries(layer.manual)) {
    if (!(setting in layer.maskedByBoth)) result[setting] = value
  }
}

export function effectiveValues(session: SimulatorSession, catalog: Catalog, eye: Eye): Record<string, Value> {
  const result = Object.fromEntries(catalog.groups.flatMap(group => group.settings.map(setting => [setting.id, setting.default])))
  applyLayer(result, session.layers.both, catalog, 'both')
  applyLayer(result, session.layers[eye], catalog, eye)
  return result
}

export function editableValues(session: SimulatorSession, catalog: Catalog): Record<string, Value> {
  if (session.eyeMode !== 'both') return effectiveValues(session, catalog, session.eyeMode)
  const result: Record<string, Value> = Object.fromEntries(
    catalog.groups.flatMap(group => group.settings.map(setting => [setting.id, setting.default])),
  )
  applyLayer(result, session.layers.both, catalog, 'both')
  return result
}

export function sourceArticle(session: SimulatorSession, catalog: Catalog, settingId: string): string | undefined {
  const target = session.eyeMode
  const layer = session.layers[target]
  if (settingId in layer.manual) return undefined
  const active = new Set(layerPresets(layer, catalog))
  const owner = catalog.presets.find(item => active.has(item.id) && settingId in presetValues(item, target))
  return owner && catalog.articles.find(article =>
    article.demonstrations.some(item => item.presets.includes(owner.id)),
  )?.id
}

File: vss-ios/App/SimulatorSession.swift
import Foundation

enum EyeMode: String, CaseIterable, Hashable { case left, both, right }

struct SessionLayer: Equatable {
    var selectedDemonstrations: [String: String] = [:]
    var manual: [String: JSONValue] = [:]
    var maskedFallback: [String: JSONValue] = [:]
    var maskedByBoth = Set<String>()
}

struct SimulatorSession: Equatable {
    var eyeMode: EyeMode = .left
    var layers: [EyeMode: SessionLayer] = Dictionary(uniqueKeysWithValues: EyeMode.allCases.map { ($0, SessionLayer()) })

    func withEyeMode(_ mode: EyeMode) -> SimulatorSession {
        var next = self; next.eyeMode = mode; return next
    }
    func currentLayer() -> SessionLayer { layer(eyeMode) }
    func selectedDemonstration(articleID: String) -> String? {
        if eyeMode != .both { return currentLayer().selectedDemonstrations[articleID] }
        return layer(.both).selectedDemonstrations[articleID]
            ?? layer(.left).selectedDemonstrations[articleID]
            ?? layer(.right).selectedDemonstrations[articleID]
    }

    func activePresets(in catalog: Catalog) -> Set<String> {
        if eyeMode != .both { return presets(in: currentLayer(), catalog: catalog) }
        let shared = presets(in: layer(.both), catalog: catalog)
        let intrinsic = presets(in: layer(.left), catalog: catalog).union(presets(in: layer(.right), catalog: catalog))
            .filter { catalog.preset($0)?.isIntrinsic == true }
        return shared.union(intrinsic)
    }

    func selectDemonstration(in catalog: Catalog, articleID: String, demonstrationID: String) -> SimulatorSession {
        guard let article = catalog.articles.first(where: { $0.id == articleID }),
              let selected = article.demonstrations.first(where: { $0.id == demonstrationID }) else { return self }
        let selectedPresets = selected.presets.compactMap(catalog.preset)
        let intrinsic = selectedPresets.contains { $0.isIntrinsic }
        let targets = intrinsic ? [EyeMode.left, .right].filter { target in selectedPresets.contains { !$0.values(for: target).isEmpty } } : [eyeMode]
        let enabled = !targets.allSatisfy { layer($0).selectedDemonstrations[articleID] == demonstrationID }
        var next = self
        for target in targets {
            next = next.withLayer(target, transition(layer: next.layer(target), catalog: catalog, target: target, articleID: articleID, selected: selected, enabled: enabled))
            if target != .both && enabled { next = next.mask(targets: [target], settings: demonstrationValues(selected, catalog: catalog, target: target).keys, masked: false) }
        }
        if targets.contains(.both) {
            let changed = Set(demonstrationValues(selected, catalog: catalog, target: .both).keys)
            next = enabled
                ? next.mask(targets: [.left, .right], settings: changed, masked: true)
                : next.mask(targets: [.left, .right], settings: changed.subtracting(next.sharedSettings(in: catalog)), masked: false)
        }
        if intrinsic && enabled { next.eyeMode = .both }
        return next
    }

    func edit(_ settingID: String, value: JSONValue) -> SimulatorSession {
        var targetLayer = currentLayer(); targetLayer.manual[settingID] = value
        var next = withLayer(eyeMode, targetLayer)
        return next.mask(targets: eyeMode == .both ? [.left, .right] : [eyeMode], settings: [settingID], masked: eyeMode == .both)
    }

    func reset(_ settingID: String, in catalog: Catalog? = nil) -> SimulatorSession {
        var targetLayer = currentLayer(); targetLayer.manual.removeValue(forKey: settingID)
        var next = withLayer(eyeMode, targetLayer)
        if eyeMode == .both && (catalog == nil || !next.sharedSettings(in: catalog!).contains(settingID)) {
            next = next.mask(targets: [.left, .right], settings: [settingID], masked: false)
        }
        return next
    }

    func effectiveValues(in catalog: Catalog, for eye: EyeMode) -> [String: JSONValue] {
        precondition(eye != .both)
        var result = Dictionary(uniqueKeysWithValues: catalog.groups.flatMap(\.settings).map { ($0.id, $0.default) })
        apply(layer(.both), to: &result, catalog: catalog, target: .both)
        apply(layer(eye), to: &result, catalog: catalog, target: eye)
        return result
    }

    func editableValues(in catalog: Catalog) -> [String: JSONValue] {
        if eyeMode != .both { return effectiveValues(in: catalog, for: eyeMode) }
        var result = Dictionary(uniqueKeysWithValues: catalog.groups.flatMap(\.settings).map { ($0.id, $0.default) })
        apply(layer(.both), to: &result, catalog: catalog, target: .both)
        return result
    }

    func sourceArticleID(in catalog: Catalog, settingID: String) -> String? {
        let targetLayer = currentLayer()
        guard targetLayer.manual[settingID] == nil,
              let owner = catalog.presets.first(where: { presets(in: targetLayer, catalog: catalog).contains($0.id) && $0.values(for: eyeMode)[settingID] != nil }) else { return nil }
        return catalog.articles.first { $0.demonstrations.contains { $0.presets.contains(owner.id) } }?.id
    }

    private func layer(_ mode: EyeMode) -> SessionLayer { layers[mode] ?? SessionLayer() }
    private func withLayer(_ mode: EyeMode, _ layer: SessionLayer) -> SimulatorSession { var next = self; next.layers[mode] = layer; return next }
    private func mask<S: Sequence>(targets: [EyeMode], settings: S, masked: Bool) -> SimulatorSession where S.Element == String {
        let settings = Set(settings); var next = self
        for target in targets { var value = next.layer(target); if masked { value.maskedByBoth.formUnion(settings) } else { value.maskedByBoth.subtract(settings) }; next.layers[target] = value }
        return next
    }
    private func sharedSettings(in catalog: Catalog) -> Set<String> {
        Set(layer(.both).manual.keys).union(settings(for: presets(in: layer(.both), catalog: catalog), catalog: catalog, target: .both))
    }
}

private extension Catalog {
    func preset(_ id: String) -> CatalogPreset? { presets.first { $0.id == id } }
}
private extension CatalogPreset {
    var isIntrinsic: Bool { !left.isEmpty || !right.isEmpty }
    func values(for target: EyeMode) -> [String: JSONValue] {
        if !both.isEmpty { return both }
        switch target { case .left: return left; case .right: return right; case .both: return [:] }
    }
}
private func presets(in layer: SessionLayer, catalog: Catalog) -> Set<String> {
    Set(catalog.articles.flatMap { article in article.demonstrations.first { $0.id == layer.selectedDemonstrations[article.id] }?.presets ?? [] })
}
private func demonstrationValues(_ demo: CatalogDemonstration, catalog: Catalog, target: EyeMode) -> [String: JSONValue] {
    demo.presets.compactMap(catalog.preset).reduce(into: [:]) { result, preset in result.merge(preset.values(for: target)) { _, new in new } }
}
private func settings(for presets: Set<String>, catalog: Catalog, target: EyeMode) -> Set<String> {
    Set(presets.compactMap(catalog.preset).flatMap { $0.values(for: target).keys })
}
private func transition(layer: SessionLayer, catalog: Catalog, target: EyeMode, articleID: String, selected: CatalogDemonstration, enabled: Bool) -> SessionLayer {
    let before = presets(in: layer, catalog: catalog); var selections = layer.selectedDemonstrations
    if !enabled { selections.removeValue(forKey: articleID) } else {
        let values = demonstrationValues(selected, catalog: catalog, target: target)
        for article in catalog.articles where article.id != articleID {
            guard let id = selections[article.id], let other = article.demonstrations.first(where: { $0.id == id }) else { continue }
            if demonstrationValues(other, catalog: catalog, target: target).contains(where: { setting, value in values[setting].map { $0 != value } ?? false }) { selections.removeValue(forKey: article.id) }
        }
        selections[articleID] = selected.id
    }
    var changed = layer; changed.selectedDemonstrations = selections
    return cascade(catalog: catalog, target: target, before: before, after: presets(in: changed, catalog: catalog), layer: changed)
}
private func cascade(catalog: Catalog, target: EyeMode, before: Set<String>, after: Set<String>, layer: SessionLayer) -> SessionLayer {
    let beforeSettings = settings(for: before, catalog: catalog, target: target), afterSettings = settings(for: after, catalog: catalog, target: target)
    var next = layer
    for setting in beforeSettings.union(afterSettings) {
        if afterSettings.contains(setting) { if let value = next.manual.removeValue(forKey: setting) { next.maskedFallback[setting] = value } }
        else if beforeSettings.contains(setting) { if next.manual[setting] != nil { next.maskedFallback.removeValue(forKey: setting) } else if let value = next.maskedFallback.removeValue(forKey: setting) { next.manual[setting] = value } }
    }
    return next
}
private func apply(_ layer: SessionLayer, to result: inout [String: JSONValue], catalog: Catalog, target: EyeMode) {
    for preset in catalog.presets where presets(in: layer, catalog: catalog).contains(preset.id) { result.merge(preset.values(for: target).filter { !layer.maskedByBoth.contains($0.key) }) { _, new in new } }
    result.merge(layer.manual.filter { !layer.maskedByBoth.contains($0.key) }) { _, new in new }
}
